doc/design/record_fetcher_proposal.md

```diff
@@ -0,0 +1,178 @@
+# Raw Record Representation Proposal
+
+## BLUF
+
+Upload fetched documents, having performed the minimal amount of processing. Define a manifest expressing the
+association between those document URIs and basic metadata (ID, region and record type, timestamp of collection, etc.).
+Sign that manifest and upload it. Now a logical record's components may be derived from the manifest and the now-stored
+documents the manifest describes.
+
+## Objective
+
+Define a system for tracking the associations between the various crude scrapings that describe a record. How documents
+are fetched and how software is deployed are both directly relevant to this space, but are out of scope for this
+document.
+
+## Background
+
+Record collection like we're trying to do is a dumpster fire of inconsistency. In some cases the fields we want are in a
+single document, but frequently the fields we require for a single logical record span multiple documents. Rather than
+having every scraper developer wrestle with this complexity, present design discussions are leaning toward separate
+source fetching and field extraction stages. This allows us to build a corpus of crude source material which may be more
+easily refined offline, with greatly reduced risk of the extraction process tainting the original data.
+
+Separately, there is a large concern surrounding tracking of data lineage to prevent (or at least reliably detect)
+accidental or intentional tampering of data. Offline processing of the raw source data makes it possible to cheaply
+iterate on extraction tooling and forward propagate the results, while also being able to compare against previous
+extractions, without the need for querying a service with questionable availability and durability.
+
+## Requirements
+
+We need a way to store raw information pertaining to logical records. Given logical records can span documents, we also
+need to store a representation (a manifest) of those relationships. Given records are capable of changing over time, we
+need to be able to track distinct snapshots over time. Additionally, we need a service which will take this information
+and persist it in a meaningful way. Analysis of this information is not in scope.
+
+### Manifests
+
+A manifest needs to be able to represent the raw materials associated with a record. The minimal identifiers for the
+record itself are:
+* a unique ID
+* which region and record type the manifest represents (e.g., "USA/FL/Bay/Court")
+* some sort of case identifier
```

Comment:
I'd personally view the case identifier as part of the records "fields" that will be extracted later based on the type of the record.

If we store different kinds of records, case identifier isn't universal.  Unique ID, timestamps, origin, PDAP URI, etc are all universal across any kind of record type

Reply:
Could you expand upon this some?

If I'm understanding you correctly, we'd analyze the record (post-fetching, of course) and then derive some sort of record identifier based upon that?

Reply:
I think what he was saying is that an attribute like this may be difficult at the moment to proclaim that it could be at a high level like this in the manifest?

His suggestion may have been slightly off base for what you were intending the text to contain below line 39/40.

Reply:
Hmmm. I think I understand this better now. Thanks for the help!

So I don't think I care about the actual case number for my imaginary concerns, it's more that I want any number that can be used to link two collected records together. If we fetch the same content twice and the document digests don't match up exactly, it'll take a lot of processing to recognize those are actually the same if we don't have some stable identifier. That said, I can totally see this being hard not just for online records, but if we try to apply the same structure to scanned documents or whatever.

In some situations, the case number is well established. For example, the Benchmark portal is case number-centric, so we *know* the number associated for those cases because we had to explicitly search for them. In those cases, we have a strong confidence in that identifier because we asked explicitly for that number and that's what the source gave us.

How would folks feel about making this optional? Or do we still think we should just not collect it? If we drop it now, we can always change out minds later (it may just require re-fetching). Thoughts?

Reply:
Some kind of record-specific identifier is a must to avoid duplication as mentioned by Josh. Case ID is the most obvious, but the absence of case IDs is inevitably going to happen on some portal(s). If we drop Case IDs, what takes its place?

In this case, it's important to consider how we should consistently create the same identifier when scraping a given court record multiple times. A crude way would be to take a bunch of fields and hash them to make an ID. Better ideas are welcomed...